src/main/java/edu/kit/datamanager/security/filter/KeycloakTokenFilter.java

```diff
@@ -0,0 +1,181 @@
+                } else {
+                    LOG.info("Invalid Request: Token is expired or tampered");
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized: Token is expired or tampered");
+                }
+            } catch (BadJOSEException | IOException | MalformedJwtException e) {
+                LOG.error("Failed to validate JWT.", e);
+            }
+
+        } else {
+            LOG.info("Authorization Token not being sent in Headers: " + token);
+        }
+
+        LOG.trace("Trying alternate authentication via JWT.");
+
+        filterChain.doFilter(request, response);
+    }
+
+    private boolean attemptLocalAuthentication(HttpServletRequest request, HttpServletResponse response, String token) {
+        boolean authenticated = false;
+        try {
+            Jws<Claims> claimsJws = tokenValidator.getJwsClaims(token);
+            Set<Map.Entry<String, Object>> claims = claimsJws.getBody().entrySet();
+            Map<String, Object> claimMap = new HashMap<>();
+            claims.forEach((entry) -> {
+                claimMap.put(entry.getKey(), entry.getValue());
+            });
+            jwToken = JwtAuthenticationToken.factoryToken(token, claimMap);
+
+            if (jwToken instanceof JwtServiceToken && ((JwtServiceToken) jwToken).getSources() != null) {
+                JwtServiceToken serviceToken = (JwtServiceToken) jwToken;
+                LOG.debug("Performing source check for JWToken for service {} and sources {}.", serviceToken.getPrincipal(), Arrays.asList(serviceToken.getSources()));
+                String remoteAddr = request.getRemoteAddr();
+                LOG.debug("Trying to match remote address {} with at least one allowed source.", remoteAddr);
+                boolean matchFound = false;
+                for (String source : serviceToken.getSources()) {
+                    if (NetworkUtils.matches(remoteAddr, source)) {
+                        matchFound = true;
+                        break;
+                    }
+                }
+                if (!matchFound) {
+                    LOG.warn("Invalid request from remote address {} to service {} found. Request denied.", remoteAddr, serviceToken.getPrincipal());
+                    throw new InvalidAuthenticationException("You are not allowed to authenticate using the provided token from your current location.");
+                }
+            }
+            LOG.trace("JWT validation finished. Checking result.");
+            if (jwToken != null && SecurityContextHolder.getContext().getAuthentication() == null) {
+                LOG.info("Authenticated username: {}", jwToken.getPrincipal());
+                setContext(request, jwToken);
+            } else {
+                try {
+                    LOG.info("Invalid Request: Token is expired or tampered");
+                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized: Token is expired or tampered");
```

Comment:
In this case method is embedded in a try-catch block!? Why handle this in different ways?

Reply:
This should not be required any longer.